pkg/gitproviders/provider_user_test.go

```diff
@@ -0,0 +1,314 @@
+			Expect(err).ToNot(HaveOccurred())
+			Expect(res).To(BeTrue())
+		})
+	})
+
+	Describe("UploadDeployKey", func() {
+		var deployKeyClient *fakegitprovider.DeployKeyClient
+
+		BeforeEach(func() {
+			deployKeyClient = &fakegitprovider.DeployKeyClient{}
+			userRepo.DeployKeysReturns(deployKeyClient)
+		})
+
+		It("return error when repo doest exist", func() {
+			userRepoClient.GetReturns(nil, gitprovider.ErrNotFound)
+
+			err := userProvider.UploadDeployKey("owner", "repo-name", []byte("my-key"))
+			Expect(err.Error()).Should(ContainSubstring("error getting user repo reference for owner"))
+		})
+
+		It("returns error when can't create the key", func() {
+			deployKeyClient.CreateReturns(nil, errors.New("random error"))
+
+			err := userProvider.UploadDeployKey("owner", "repo-name", []byte("my-key"))
+			Expect(err.Error()).Should(ContainSubstring("error uploading deploy key"))
+		})
+
+		It("creates the deploy key", func() {
+			deployKeyClient.CreateReturns(nil, nil)
+			deployKeyClient.GetReturns(nil, nil)
+
+			err := userProvider.UploadDeployKey("owner", "repo-name", []byte("my-key"))
+			Expect(err).ToNot(HaveOccurred())
+		})
+	})
+
+	Describe("GetDefaultBranch", func() {
+		It("returns error when cant extract owner from url", func() {
+			_, err := userProvider.GetDefaultBranch("bad-url")
+			Expect(err.Error()).Should(ContainSubstring("could get provider name from URL"))
+		})
+
+		It("returns error when can't get branch", func() {
+			userRepoClient.GetReturns(nil, gitprovider.ErrNotFound)
+
+			_, err := userProvider.GetDefaultBranch("http://github.com/owner/repo")
+			Expect(err.Error()).Should(ContainSubstring("error getting user repository"))
+		})
+
+		It("returns repo default branch", func() {
+			userRepo.GetReturns(gitprovider.RepositoryInfo{DefaultBranch: gitprovider.StringVar("my-branch")})
+
+			branch, err := userProvider.GetDefaultBranch("http://github.com/owner/repo")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(branch).To(Equal("my-branch"))
+		})
+	})
+
+	Describe("GetRepoVisibility", func() {
+		It("returns error when cant extract owner from url", func() {
+			_, err := userProvider.GetRepoVisibility("bad-url")
+			Expect(err.Error()).Should(ContainSubstring("could get provider name from URL"))
+		})
+
+		It("returns error when can't get branch", func() {
+			userRepoClient.GetReturns(nil, gitprovider.ErrNotFound)
+
+			_, err := userProvider.GetRepoVisibility("http://github.com/owner/repo")
+			Expect(err.Error()).Should(ContainSubstring("error getting user repository"))
+		})
+
+		It("returns repo default branch", func() {
+			visibility := gitprovider.RepositoryVisibilityVar(gitprovider.RepositoryVisibilityPrivate)
+			userRepo.GetReturns(gitprovider.RepositoryInfo{Visibility: visibility})
+
+			vis, err := userProvider.GetRepoVisibility("http://github.com/owner/repo")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vis).To(Equal(visibility))
+		})
+	})
+
+	Describe("CreatePullRequest", func() {
```

Comment:
Seeing all these cases duplicated makes me wonder if we should do some sort of table-like test for both implementations. Probably outside the scope for the PR.

Reply:
yes, I don't like either, I'll spend some time on this to see if I can come with a good solution for it.